src/compiler/reflection.jl

```diff
@@ -64,35 +45,30 @@ function code_sass(io::IO, job::CompilerJob; raw::Bool=false)
         return
     end
 
-    compiled = compile(job)
+    cfg = CUPTI.CallbackConfig([CUPTI.CUPTI_CB_DOMAIN_RESOURCE]) do domain, id, data
+        id == CUPTI.CUPTI_CBID_RESOURCE_MODULE_LOADED || return
+        resourceDescriptor =
+            unsafe_load(convert(Ptr{CUpti_ModuleResourceData}, data.resourceDescriptor))
+        cubin = unsafe_wrap(Vector{Cchar}, pointer(resourceDescriptor.pCubin),
+                            resourceDescriptor.cubinSize)
+        disassemble_cubin(io, cubin; raw)
+    end
 
-    cubin = Ref{Any}()
-    callback = @cfunction(code_sass_callback, Cvoid,
-                          (Ptr{Cvoid}, CUpti_CallbackDomain, CUpti_CallbackId, Ptr{Cvoid}))
-
-    # JIT compile and capture the generated object file
-    lock(cupti_lock) do
-        subscriber_ref = Ref{CUpti_SubscriberHandle}()
-        res = CUPTI.unsafe_cuptiSubscribe(subscriber_ref, callback, Base.pointer_from_objref(cubin))
-        if res === CUPTI.CUPTI_ERROR_INSUFFICIENT_PRIVILEGES
-            error("""Insufficient privilege: You don't have permissions to profile GPU code, which is required for `code_sass`.
-                    Get administrative privileges or allow all users to profile: https://developer.nvidia.com/ERR_NVGPUCTRPERM#SolnAdminTag""")
-        elseif res != CUPTI.CUPTI_SUCCESS
-            throw(CUPTIError(res))
-        end
-        subscriber = subscriber_ref[]
-        try
-            CUPTI.cuptiEnableDomain(1, subscriber, CUPTI.CUPTI_CB_DOMAIN_RESOURCE)
-            link(job, compiled)
-        finally
-            CUPTI.cuptiUnsubscribe(subscriber)
-        end
+    compiled = compile(job)
+    CUPTI.enable!(cfg)
+    try
+        link(job, compiled)
+    finally
+        CUPTI.disable!(cfg)
     end
+  
```

Comment:
I prefer making the lifetime requirement explicit. `CUPTI.enable!` leaks the cfg pointer to C and `disable!` ends the pointer leakage. Or we could document that enable!/disable! need to be explicitly matched and a `CUPTI.enable!(cfg)` on it's own is disallowed.

```suggestion
    GC.@preserve cfg begin
        CUPTI.enable!(cfg)
        try
            link(job, compiled)
        finally
            CUPTI.disable!(cfg)
        end
    end
```

Reply:
Yeah I wasn't sure whether that's needed given Julia's semantics. The `cfg` object needs to be kept alive for the `disable!` call, but I'm not sure what SROA does with that.

Reply:
I hate reasoning about the semantics of optimizations xD That's why I prefer explicit managing of lifetimes in those situations.

1. `pointer_from_objref` should either be inside a `GC.@preserve` or after a manual escape to a global.
2. SROA should only be allowed to touch mutable objects iff there are no escapes
3. For the higher-level optimizer  `pointer_from_objref` or the corresponding `jl_value_ptr` have no additional semantics
    (I think that means `jl_value_ptr` counts as a momentary escape and thus SROA should not happen in the higher-level)
4. In alloc-opt I think `jl_value_ptr` is an address escape thus preventing heap-to-stack

So I think the code as written now is safe enough, but wouldn't be if someone starts moving `CUPTI.disable!(cfg)` or commenting it out.

Reply:
Exposing `GC.@preserve` is just so ugly. Maybe a macro, or yet another closure?